internal/ls/autoimport/registry.go

```diff
@@ -0,0 +1,1030 @@
+
+	var exportsMu sync.Mutex
+	exports := make(map[tspath.Path][]*Export)
+	ambientModuleNames := make(map[string][]string)
+
+	var entrypointsMu sync.Mutex
+	var entrypoints []*module.ResolvedEntrypoints
+	var skippedEntrypointsCount int32
+	var combinedStats extractorStats
+	var possibleFailedAmbientModuleLookupTargets collections.SyncSet[string]
+	var possibleFailedAmbientModuleLookupSources collections.SyncMap[tspath.Path, *failedAmbientModuleLookupSource]
+
+	createAliasResolver := func(packageName string, entrypoints []*module.ResolvedEntrypoint) *aliasResolver {
+		rootFiles := make([]*ast.SourceFile, len(entrypoints))
+		var wg sync.WaitGroup
+		for i, entrypoint := range entrypoints {
+			wg.Go(func() {
+				file := b.host.GetSourceFile(entrypoint.ResolvedFileName, b.base.toPath(entrypoint.ResolvedFileName))
+				binder.BindSourceFile(file)
+				rootFiles[i] = file
+			})
+		}
+		wg.Wait()
+
+		rootFiles = slices.DeleteFunc(rootFiles, func(f *ast.SourceFile) bool {
+			return f == nil
+		})
+
+		return newAliasResolver(rootFiles, b.host, b.resolver, b.base.toPath, func(source ast.HasFileName, moduleName string) {
+			possibleFailedAmbientModuleLookupTargets.Add(moduleName)
+			possibleFailedAmbientModuleLookupSources.LoadOrStore(source.Path(), &failedAmbientModuleLookupSource{
+				fileName: source.FileName(),
+			})
+		})
+	}
+
+	indexStart := time.Now()
+	var wg sync.WaitGroup
+	for packageName := range packageNames.Keys() {
+		wg.Go(func() {
+			if ctx.Err() != nil {
+				return
+			}
+
+			typesPackageName := module.GetTypesPackageName(packageName)
+			var packageJson *packagejson.InfoCacheEntry
+			packageJson = b.host.GetPackageJson(tspath.CombinePaths(dirName, "node_modules", packageName, "package.json"))
+			if !packageJson.DirectoryExists {
+				packageJson = b.host.GetPackageJson(tspath.CombinePaths(dirName, "node_modules", typesPackageName, "package.json"))
+			}
+			packageEntrypoints := b.resolver.GetEntrypointsFromPackageJsonInfo(packageJson, packageName)
+			if packageEntrypoints == nil {
+				return
+			}
+			if len(fileExcludePatterns) > 0 {
+				count := int32(len(packageEntrypoints.Entrypoints))
+				packageEntrypoints.Entrypoints = slices.DeleteFunc(packageEntrypoints.Entrypoints, func(entrypoint *module.ResolvedEntrypoint) bool {
+					for _, excludePattern := range fileExcludePatterns {
+						if matched, _ := excludePattern.MatchString(entrypoint.ResolvedFileName); matched {
+							return true
+						}
+					}
+					return false
+				})
+				atomic.AddInt32(&skippedEntrypointsCount, count-int32(len(packageEntrypoints.Entrypoints)))
+			}
+			if len(packageEntrypoints.Entrypoints) == 0 {
+				return
+			}
+
+			entrypointsMu.Lock()
+			entrypoints = append(entrypoints, packageEntrypoints)
+			entrypointsMu.Unlock()
+
+			aliasResolver := createAliasResolver(packageName, packageEntrypoints.Entrypoints)
+			checker, _ := checker.NewChecker(aliasResolver)
```

Comment:
How often does this happen? Creating checkers is quite expensive, technically, so I assume not often